### ISSUE TITLE
bug

### ISSUE THREAD

**Issue by cymdhx** [ChristophReich1996/Involution#3]

**Comment by cymdhx** [comment]
when I  use involution replace conv. I meet this problem
![image](https://user-images.githubusercontent.com/73769530/113833101-d7322b80-97bb-11eb-9d50-23dfe02fe06c.png)

**Comment by ChristophReich1996** [comment]
Hi cymdhx, regrettably I can not track this issue down by the raw error message. Please provide more information, like the parameters of the involutions and the shape of the input tensor.

**Comment by cymdhx** [comment]
![image](https://user-images.githubusercontent.com/73769530/113866977-8bde4400-97e0-11eb-9e00-1162a5e5697b.png)
我是把yolov4的PANet部分的卷积替换成involution，其他的没有变化，当我使用基础卷积的时候可以正常，但是用了involution就出现了这个问题。

**Comment by ChristophReich1996** [comment]
Sorry, my Chinese is pretty bad, could you write in English? I need the **parameters** (in_channels, out_channels, sigma_mapping, kernel_size, stride, groups, reduce_ratio, dilation, and padding) and the **input shape** of the involution where the error occurs!

**Comment by cymdhx** [comment]
> ![image](https://user-images.githubusercontent.com/73769530/113866977-8bde4400-97e0-11eb-9e00-1162a5e5697b.png)
> 我是把yolov4的PANet部分的卷积替换成involution，其他的没有变化，当我使用基础卷积的时候可以正常，但是用了involution就出现了这个问题。

I'm replacing the convolution of the panet part of yolov4 with involution, and the others are unchanged. When I use the basic convolution, it can work normally, but this problem occurs when I use the evolution
The initial one is like this：
![image](https://user-images.githubusercontent.com/73769530/113972424-2edc9f80-986d-11eb-8f1f-4ae79bdcdecb.png)
Then I changed to this：
![image](https://user-images.githubusercontent.com/73769530/113973603-27b69100-986f-11eb-955a-eede8d1bbfd6.png)

This line reports mistakes
![image](https://user-images.githubusercontent.com/73769530/113972854-f4273700-986d-11eb-9f59-cc6cfd0f967e.png)
![image](https://user-images.githubusercontent.com/73769530/113973880-9b589e00-986f-11eb-8a27-59caa1549331.png)





you can look in https://github.com/argusswift/YOLOv4-pytorch/blob/master/model/YOLOv4.py

parameters like this
![image](https://user-images.githubusercontent.com/73769530/113988808-6a359900-9882-11eb-9a9c-a6e185d17bf2.png)


class PANet(nn.Module):
    def __init__(self, feature_channels):
        super(PANet, self).__init__()

        self.feature_transform3 = Conv(
            feature_channels[0], feature_channels[0] // 2, 1
        )
        self.feature_transform4 = Conv(
            feature_channels[1], feature_channels[1] // 2, 1
        )

        self.resample5_4 = Upsample(
            feature_channels[2] // 2, feature_channels[1] // 2
        )
        self.resample4_3 = Upsample(
            feature_channels[1] // 2, feature_channels[0] // 2
        )
        self.resample3_4 = Downsample(
            feature_channels[0] // 2, feature_channels[1] // 2
        )
        self.resample4_5 = Downsample(
            feature_channels[1] // 2, feature_channels[2] // 2
        )

        self.downstream_conv5 = nn.Sequential(
            Conv(feature_channels[2] * 2, feature_channels[2] // 2, 1),
            Conv(feature_channels[2] // 2, feature_channels[2], 3),
            Conv(feature_channels[2], feature_channels[2] // 2, 1),
        )
        self.downstream_conv4 = nn.Sequential(
            Conv(feature_channels[1], feature_channels[1] // 2, 1),
            Conv(feature_channels[1] // 2, feature_channels[1], 3),
            Conv(feature_channels[1], feature_channels[1] // 2, 1),
            Conv(feature_channels[1] // 2, feature_channels[1], 3),
            Conv(feature_channels[1], feature_channels[1] // 2, 1),
        )
        self.downstream_conv3 = nn.Sequential(
            Conv(feature_channels[0], feature_channels[0] // 2, 1),
            Conv(feature_channels[0] // 2, feature_channels[0], 3),
            Conv(feature_channels[0], feature_channels[0] // 2, 1),
            Conv(feature_channels[0] // 2, feature_channels[0], 3),
            Conv(feature_channels[0], feature_channels[0] // 2, 1),
        )

        self.upstream_conv4 = nn.Sequential(
            Conv(feature_channels[1], feature_channels[1] // 2, 1),
            Conv(feature_channels[1] // 2, feature_channels[1], 3),
            Conv(feature_channels[1], feature_channels[1] // 2, 1),
            Conv(feature_channels[1] // 2, feature_channels[1], 3),
            Conv(feature_channels[1], feature_channels[1] // 2, 1),
        )
        self.upstream_conv5 = nn.Sequential(
            Conv(feature_channels[2], feature_channels[2] // 2, 1),
            Conv(feature_channels[2] // 2, feature_channels[2], 3),
            Conv(feature_channels[2], feature_channels[2] // 2, 1),
            Conv(feature_channels[2] // 2, feature_channels[2], 3),
            Conv(feature_channels[2], feature_channels[2] // 2, 1),
        )
        self.__initialize_weights()

**Comment by cymdhx** [comment]
[https://github.com/argusswift/YOLOv4-pytorch/blob/master/model/YOLOv4.py](url)

**Comment by cymdhx** [comment]
> > ![image](https://user-images.githubusercontent.com/73769530/113866977-8bde4400-97e0-11eb-9e00-1162a5e5697b.png)
> > 我是把yolov4的PANet部分的卷积替换成involution，其他的没有变化，当我使用基础卷积的时候可以正常，但是用了involution就出现了这个问题。
> 
> I'm replacing the convolution of the panet part of yolov4 with involution, and the others are unchanged. When I use the basic convolution, it can work normally, but this problem occurs when I use the evolution
> The initial one is like this：
> ![image](https://user-images.githubusercontent.com/73769530/113972424-2edc9f80-986d-11eb-8f1f-4ae79bdcdecb.png)
> Then I changed to this：
> ![image](https://user-images.githubusercontent.com/73769530/113972453-38660780-986d-11eb-820f-794c06803cb7.png)
> This line reports mistakes
> ![image](https://user-images.githubusercontent.com/73769530/113972854-f4273700-986d-11eb-9f59-cc6cfd0f967e.png)
> 
> you can look code in https://github.com/argusswift/YOLOv4-pytorch/blob/master/model/YOLOv4.py

**Comment by ChristophReich1996** [comment]
You utilized the involution for downsampling as a replacement for a strided convolution. This results in the mismatch of the size. I would suggest using a standard convolution for downsampling or a pooling layer followed by an involution.

**Comment by cymdhx** [comment]
> You utilized the involution for downsampling as a replacement for a strided convolution. This results in the mismatch of the size. I would suggest using a standard convolution for downsampling or a pooling layer followed by an involution.

Thank you so much. I solved the problem in your way. But there is a new problem. Loss will be Nan in a moment
![image](https://user-images.githubusercontent.com/73769530/114132453-f5706680-9936-11eb-9b9a-6e515995964c.png)

Is this gradient vanishing or gradient explosion.It doesn't seem to help if I turn down LR

**Comment by cymdhx** [comment]
> > You utilized the involution for downsampling as a replacement for a strided convolution. This results in the mismatch of the size. I would suggest using a standard convolution for downsampling or a pooling layer followed by an involution.
> 
> Thank you so much. I solved the problem in your way. But there is a new problem. Loss will be Nan in a moment
> ![image](https://user-images.githubusercontent.com/73769530/114132453-f5706680-9936-11eb-9b9a-6e515995964c.png)
> 
> Is this gradient vanishing or gradient explosion.It doesn't seem to help if I turn down LR

If I delete batchnorm2d after involution , Nan almost not appear. Do you know why

**Comment by ChristophReich1996** [comment]
I guess, this is due to exploding gradients or activation. This needs to be investigated further. In case of exploding gradients, gradient clipping could help. When this issue is caused by exploding activation try setting affine and track_running_stats in all batch normalization layers to False.

**Comment by cymdhx** [comment]
> I guess, this is due to exploding gradients or activation. This needs to be investigated further. In case of exploding gradients, gradient clipping could help. When this issue is caused by exploding activation try setting affine and track_running_stats in all batch normalization layers to False.

Thank you. I deleted the BN(BatchNorm2d) in the above picture, and then greatly reduced the probability of nan，Only occasionally。Do you know why.sorry to bother you.
![image](https://user-images.githubusercontent.com/73769530/114190138-8b2ee480-997d-11eb-8c27-f73b2df586d2.png)

**Comment by ChristophReich1996** [comment]
This is kinda odd, because if the problem caused by very large activation (which is for me more obvious than exploding gradients) batch normalization should prevent this problem. But maybe I'am wrong. However, this should be investigated further by analyzing the distribution of the activations and gradients during training. Maybe the author's have a more sophisticated explanation.

**Comment by cymdhx** [comment]
> This is kinda odd, because if the problem caused by very large activation (which is for me more obvious than exploding gradients) batch normalization should prevent this problem. But maybe I'am wrong. However, this should be investigated further by analyzing the distribution of the activations and gradients during training. Maybe the author's have a more sophisticated explanation.

thank you so much
